source/testing/index.rst

```diff
@@ -0,0 +1,499 @@
+  and `video <http://www.youtube.com/watch?v=ickNQcNXiS4>`_. Unfortunately,
+  the first part (slides 2-12) is
+  out-of-date with behavior of current versions of Django (1.9 as I write this), so
+  skip ahead to slide 13 and continue from there.
+
+How to test Django template tags
+++++++++++++++++++++++++++++++++
+
+Testing template tags is an often overlooked and difficult aspect of
+testing in Django apps, but if you break up the logic right, you can
+actually test your custom tags easily.
+
+As an example, testing separately the parsing of a tag::
+
+    @patch('django.template.Variable')
+    def test_tag(self, Variable):
+        parser = Mock()
+        token = Mock(methods=['split_contents'])
+
+        token.split_contents.return_value = (
+            'link_to_email', 'bob@company.com')
+        node = LinkToEmail.tag(parser, token)
+
+        self.assertEqual(node.email, Variable.return_value)
+        assert not node.obfuscate
+
+And the rendering of the tag as a second test::
+
+    class ContextMock(dict):
+        autoescape = object()
+
+    @patch('django.template.loader.get_template')
+    @patch('django.template.Context')
+    def test_link_to_email_render(self, get_template, Context):
+        template = get_template.return_value
+        node = LinkToEmail(obfuscate=False, email=Mock())
+        node.email.resolve.return_value = 'bob@company.com'
+
+        context = ContextMock({
+            'email': Mock(),
+            'obfuscate': Mock(),
+        })
+
+        node.render(context)
+        template.render.assert_called_with(context.return_value)
+
+        args, kwargs = context.call_args
+        assert kwargs['autoescape'] is context.autoescape
+        assert args[0]['email'] is context['email']
+        assert args[0]['obfuscate'] is context['obfuscate']
+
+
+References:
+
+* `How To Test Django Template Tags - Part 1 <http://techblog.ironfroggy.com/2008/10/how-to-test-django-template-tags-part-1.html>`_
+* `How To Test Django Template Tags - Part 2 <http://techblog.ironfroggy.com/2008/10/how-to-test-django-template-tags-part-2.html>`_
+
+What to mock and when
++++++++++++++++++++++
+
+One of the most commonly used tools in writing Python and Django tests
+these days is the mock library, which provides an extremely simple and
+flexible way of faking parts of your code.
+
+This lets you test specific areas of an application, while assuring
+other pieces it depends on act correctly for the conditions you want
+to test.  It also lets you run tests without depending on external
+services or expensive computation that the behavior depends on
+normally.
+
+This lets your tests focus which parts of the code they actually
+cover, and a failing test is a much more narrow path to the code it
+alerts you to.
+
+Good things to mock include:
+
+* Routines which do anything with network or IO access. This also speeds up tests considerably!
+* 3rd party libraries
+* Other applications
+* Routines depending on the time of day
```

Comment:
* Ensuring that logging is called appropriately

(I've used it for that ^... not sure if it is an accepted practice)